### ISSUE TITLE
Diagnostic highlighting stuck

### ISSUE THREAD

**Issue by MariaSolOs** [neovim/neovim#24968]
### Problem

Something between https://github.com/neovim/neovim/commit/1635c9e75e21e07c4331cf983e14a11c7e09b119 and https://github.com/neovim/neovim/commit/794981d9bed7048fb3ee1ada38fcf1ebdace4c53 changed highlighting, and the highlights from LSP diagnostics remain even after fixing the error:

https://github.com/neovim/neovim/assets/62502207/8e3bd2fc-6657-46b7-9177-f7843ab90483

Notice how in the video above the highlights in `mode_component()` are messed up even after I removed the problematic code above.

Note that if I rollback to the first commit mentioned above and change _nothing_ else, I cannot reproduce this anymore.

### Steps to reproduce

1. Create a `lua` file `test.lua` with the following content:
```lua
local function get_or_create_statusline_hl(hl)
  local hl_name = "Statusline" .. hl

  if not statusline_hls[hl] then
    -- If not in the cache, create the highlight group using the icon's foreground color
    -- and the statusline's background color.
    local bg_hl = vim.api.nvim_get_hl(0, { name = "StatusLine" })
    local fg_hl = vim.api.nvim_get_hl(0, { name = hl })
    vim.api.nvim_set_hl(
      0,
      hl_name,
      { bg = ("#%06x"):format(bg_hl.bg), fg = ("#%06x"):format(fg_hl.fg) }
    )
    statusline_hls[hl] = true
  end

  return hl_name
end
```

2. Write the following `nvim` configuration in `repro.lua`:
```lua
for name, url in pairs({
  treesitter = "https://github.com/nvim-treesitter/nvim-treesitter",
  mason = "https://github.com/williamboman/mason.nvim",
  masonlsp = "https://github.com/williamboman/mason-lspconfig.nvim",
  lspconfig = "https://github.com/neovim/nvim-lspconfig",
  dracula = "https://github.com/Mofiqul/dracula.nvim",
}) do
  local install_path = vim.fn.fnamemodify("nvim_issue/" .. name, ":p")
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system({ "git", "clone", "--depth=1", url, install_path })
  end
  vim.opt.runtimepath:append(install_path)
end

require("dracula").setup()
vim.cmd.colorscheme("dracula")

require("nvim-treesitter.configs").setup({
  ensure_installed = { "lua" },
  highlight = { enable = true },
})

require("mason").setup()
require("mason-lspconfig").setup({
  ensure_installed = { "lua_ls" },
  handlers = {
    function(server)
      require("lspconfig")[server].setup({})
    end,
  },
})
```

3. Run `nvim -u repro.lua test.lua`. Below line 2, write `local`. Go back to normal mode, delete the line, and save the file. Some of the keyword highlights in the code below will still be coloured red. Sometimes it might take a couple of tries to reproduce it, but the above snippet has consistently reproduced the issue for me.

### Expected behavior

For highlights to be updated when diagnostics change.

### Neovim version (nvim -v)

v0.10.0-dev-1002+gee56daebb

### Vim (not Nvim) behaves the same?

No

### Operating system/version

Mac Ventura 13.5.1

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

xterm-kitty

### Installation

build from repo

**Comment by MariaSolOs** [comment]
@bfredl I've bisected this locally and I can trace back to https://github.com/neovim/neovim/pull/24824. Do you think there's something from that PR that could have caused this?

**Comment by MariaSolOs** [comment]
Closing this as it seems to have been fixed with #24974.
